### ISSUE TITLE
Bundled tool to convert GLTF/GLB to USDZ on Linux/OSX

### ISSUE THREAD

**Issue by cdata** [GoogleWebComponents/model-viewer#278]
_As a 3D content publisher,
when I deploy my production content,
I would like to generate USDZs from my GLTFs,
so that it is easy to support iOS users_

This issue proposes that we build and maintain a distributable, all-in-one build tool that makes it easy to generate a USDZ from a GLTF. Such a tool will allow content publishers to maintain GLTFs as their primary assets, and generate USDZs as an artifact as they go to production.

**Comment by jsantell** [comment]
Resources:

https://github.com/TimvanScherpenzeel/gltf-to-usdz-research
https://blendersushi.blogspot.com/2018/09/convert-gltf-into-usdz_3.html
https://github.com/kcoley/gltf2usd

**Comment by cwervo** [comment]
Does https://github.com/google/usd_from_gltf close this issue?

**Comment by cdata** [comment]
@cwervo it is definitely a start. I think what we should do to close out this issue is to bundle that converter as an NPM package so that web folk can trivially incorporate it into their build pipelines.

**Comment by cwervo** [comment]
Oh an implementation on npm would be great!! 👍🏼👍🏼👍🏼

**Comment by Giorat** [comment]
There's also a docker version of https://github.com/google/usd_from_gltf here: https://github.com/Giorat/docker-gltf-to-udsz

But it's really heavy in the build phase due to docker and compilation time of USDZ

**Comment by DavidPeicho** [comment]
in 2020 there is now the python [usdzconver](https://developer.apple.com/download/more/?=USDPython) script from Apple.

I tried multiple glTF and it worked pretty well. I couldnt make it work with Morph Target though.

**Comment by Giorat** [comment]
> in 2020 there is now the python [usdzconvert](https://developer.apple.com/download/more/?=USDPython) script from Apple.
> 
> I tried multiple glTF and it worked pretty well. I couldnt make it work with Morph Target though.

usdzconvert is only for Mac os? Were you able to test it on linux or mac?

**Comment by atteneder** [comment]
> > in 2020 there is now the python [usdzconvert](https://developer.apple.com/download/more/?=USDPython) script from Apple.
> > I tried multiple glTF and it worked pretty well. I couldnt make it work with Morph Target though.
> 
> usdzconvert is only for Mac os? Were you able to test it on Linux or Windows?

Yes, out of the box it's macOS only, since it contains native libraries for macOS only.

Presuming Apple didn't change anything in any dependecy lib, theoretically one could build all the native libs for other platforms like Windows and Linux and use it. But that's a bit of work (getting USD libs to compile is not straight forward).

So no, Apple's usdzconvert is no Linux solution.

**Comment by cdata** [comment]
@DavidPeicho at the time of authorship `usd_from_gltf` was considered to be faster and more faithful in converting glTF to USD compared to the tools provided by Apple. It was written by engineers working on Search who use it to automatically produce USDs from third-party glTFs displayed in search results. In the past, it has produced good results on a lot of models that appeared broken when using `usdzconvert`.

I'm not sure if those qualifications are still true today. I assume both projects have been evolving steadily over time.

**Comment by DavidPeicho** [comment]
Yes I was just bringing another alternative. I looked into the tool more deeply and indeed it looks like a lot to make it work on linux. Probably feasible, surely not easy 😢

**Comment by DavidPeicho** [comment]
To revive this a bit more, I gave a shot to [usd_from_gltf](https://github.com/google/usd_from_gltf). It's pretty good and fast. However, as the [README](https://github.com/google/usd_from_gltf/blob/master/README.md) states, they don't support Morph Targets. May not be a big deal for most users, but in my case I make heavy use of them.

I gave a shot as well to [gltf2usd](https://github.com/kcoley/gltf2usd). Worked pretty well on really complex model, but messed up the Z axis and made my material transparent (easy to fix).

**Comment by cdata** [comment]
Thanks for checking out those tools @DavidPeicho . We should probably do a survey to get a sense of the limitations of all the tools that are currently available.

**Comment by ronpadz** [comment]
It seems to me that an interim solution to this problem would be to encourage @donmccurdy & Co. to integrate usd_from_gltf into https://threejs.org/editor/ exporters.

**Comment by Weltfern** [comment]
@cdata is there any npm package or a functionality in model-viewer library that we can use to convert GLB model or GLTF file to USDZ ?

**Comment by lanxinger** [comment]
@Weltfern you can actually use apples usdz converter or Google’s gltf to usd repository to do that. You can download prebuilt USD binaries from Nvidia if you don’t want to build it yourself.

**Comment by Nnadozie** [comment]
@lanxinger apple's usdz converter only works on mac and Google’s gltf to usd install steps are daunting. There still isn't an easy to use bundled tool that's (web) accessible to non-developers or mac owners.

**Comment by donmccurdy** [comment]
Any reliable tool for creating or reading USDZ files is going to have to rely on the [USD Runtime](https://github.com/PixarAnimationStudios/USD). It doesn't work on the web, and I'm not aware of any plans to fix that. You may need to request this from Apple or from Pixar, because there's not much that web developers can do about things without it.

**Comment by elalish** [comment]
I would highly recommend this website as a solution: https://spase.io/playground
